### PR TITLE
webhook: add validating admission policy

### LINKED ISSUE #12938: Use CEL instead of webhooks
**Is this a bug report or feature request?**
* Feature Request

CEL is beta in kubernetes 1.25 and its going to be GA in 1.29 and i think it can replace webhooks and its complexity in Rook as well.

**What should the feature do:**
 using CEL for validation instead of webhook

**What is use case behind this feature:**

low maintenance and remove extra webhooks

**Environment**:

### PR DESCRIPTION
**Pull request by subhamkrai**
webhook: add validating admission policy

this commit adds validating admission policy
for cephcluster cr according to webhook rules.
Not all the webhook can be moved to validating
admission policy for example checking multus
selector validation.

**Note: Not all the Webhooks can be moved to CEL for example validating network for multus selector. And, will push changes in different commits based on local testing**

**Which issue is resolved by this Pull Request:**
Resolves #12938

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by subhamkrai** [comment]
PR is not ready for initial reviews

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 1]
```diff
@@ -534,6 +535,8 @@ const (
 )
 
 // MonSpec represents the specification of the monitor
+// +kubebuilder:validation:XValidation:message="zones must be less than or equal to count",rule="((!has(self.zones)) || (has(self.count) && size(self.zones) <= self.count))"
+// +kubebuilder:validation:XValidation:message="strectCluster zones must be less than or equal to count",rule="((!has(self.stretchCluster)) || ((size(self.stretchCluster.zones) > 0) && size(self.stretchCluster.zones) <= self.count))"
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:539]:
```suggestion
// +kubebuilder:validation:XValidation:message="stretchCluster zones must be less than or equal to count",rule="((!has(self.stretchCluster)) || ((size(self.stretchCluster.zones) > 0) && size(self.stretchCluster.zones) <= self.count))"
```

**Comment by subhamkrai** [comment]
Feel free to review PR and also PTAL at the PR description for the update. Thanks

**Review by travisn — COMMENTED** [review]
Are you planning on removing the webhook as a separate commit in this PR or a separate PR?

**Comment by subhamkrai** [comment]
> Are you planning on removing the webhook as a separate commit in this PR or a separate PR?

well, I was talking in the huddle that adding CEL takes a good amount of time to implement all the rules we have and we can't add all the webhook rules to CEL due to its own limitations. And anyway we have disabled the webhook for multiple releases so we should be free to remove the webhook and let CEL pr grow in it's own time and pace.

I think @BlaineEXE was agreeing with this if I'm not mistaken, he can correct me. @travisn wdyt?

**Comment by travisn** [comment]
> > Are you planning on removing the webhook as a separate commit in this PR or a separate PR?
> 
> well, I was talking in the huddle that adding CEL takes a good amount of time to implement all the rules we have and we can't add all the webhook rules to CEL due to its own limitations. And anyway we have disabled the webhook for multiple releases so we should be free to remove the webhook and let CEL pr grow in it's own time and pace.
> 
> I think @BlaineEXE was agreeing with this if I'm not mistaken, he can correct me. @travisn wdyt?

As a separate PR sounds good, if you could just update the title of the PR to indicate that this is just for adding CEL.

**Comment by BlaineEXE** [comment]
Since we are on the topic of the commit and pr title, it might be moer helpful to change CEL to "Validating Admission Policy," which is the name of the k8s feature. CEL is just the framework used to implement VAP.

**Comment by subhamkrai** [comment]
> Also, have you run across the CEL playground? https://playcel.undistro.io/

Yes, I have been to this site but I'm really using this or know how to use for our case.

**Comment by subhamkrai** [comment]
@travisn @BlaineEXE I'm planning to add a validating webhook policy for each CR in different PR so that it will be easier for us to review and also for me to work. This will also have advance that one validating webhook policy will settle by the next merge and not everything is broken at once. 

So I'm marking this pr to ready for review.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 2]
```diff
@@ -94,6 +94,13 @@ type CephClusterList struct {
...
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.network)))))",message="external mode enabled cluster cannot have network field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.storage)))))",message="external mode enabled cluster cannot have storage field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.disruptionManagement)))))",message="external mode enabled cluster cannot have disruptionManagement field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.mgr)))))",message="external mode enabled cluster cannot have mgr field in CR"
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:103]: Is it possible to reorder the params so that message comes first? When reading these in the future, it might be easier to read the intent first, followed by the rule definition after.

IMO, that makes it slightly easier to see at a glance what validations are implemented to see if new ones need to be added. It also makes it slightly easier to find a specific validation to edit.

```
// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.storage)))))",message="external mode enabled cluster cannot have storage field in CR"
```

```
// +kubebuilder:validation:XValidation:message="external mode enabled cluster cannot have storage field in CR",rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.storage)))))"
```

**Review by BlaineEXE — APPROVED** [review]
Approving, with a thought for potentially making things more readable when editing in the future.

**Comment by BlaineEXE** [comment]
Also, with all of the test flakes, is this PR rebased with the latest master that has some CI fixes?

**Review by travisn — CHANGES_REQUESTED** [review]
Per discussion in slack, we will need to generate a separate crds.yaml to a subdirectory that will be used for different versions of K8s. 

According to [this article](https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/), the validations graduated to beta in K8s 1.25, so I would hope that 1.25 and newer can use these. I would suggest we not enable any feature gates to get this working even in the CI. We should only recommend it be used without the feature gate enabled.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go:103]: I'm thinking of removing this validating rule since this is causing the `MultiCluster`  CI to fail because there we test external cluster and define `mon` and other fields.  Also,  it may cause issues with consumer and provider mode type clusters. @Madhu-1 any idea if this will affect you?

Comment (Madhu-1) [reply pkg/apis/ceph.rook.io/v1/types.go:103]: in provider model its a internal ceph cluster with Host Network, nothing extra

**Comment by subhamkrai** [comment]
> Also, with all of the test flakes, is this PR rebased with the latest master that has some CI fixes?

Yes, the problem is kubernetes versions

**Comment by subhamkrai** [comment]
I'm marking pr do-not-merge as I'm testing the crds for older versions of Kubernetes(less than 1.26), added separate commit for that

File: deploy/examples/pre-1.26-crds.yaml
Comment (travisn) [inline deploy/examples/pre-1.26-crds.yaml:1]: How about putting it in `deploy/examples/pre-1.25/crds.yaml`? In the past we found it is nice to keep different versions in subdirectories.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:103]: The external cluster shouldn't be defining the mons or other fields, so the rules should be fine. Note that two cephcluster CRs are created in the multicluster test, one for external and one for internal. The external one should not be defining mons or other ceph settings.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 3]
```diff
@@ -94,6 +94,13 @@ type CephClusterList struct {
 }
 
 // ClusterSpec represents the specification of Ceph Cluster
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.mon)))))",message="external mode enabled cluster cannot have mon field in CR"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:97]: I'd suggest similar wording in all the messages. Just talk about the settings instead of a "mode" in the CR
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.mon)))))",message="external clusters cannot define the mon settings"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 4]
```diff
@@ -94,6 +94,13 @@ type CephClusterList struct {
...
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.dashboard)))))",message="external mode enabled cluster cannot have dashboard field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.monitoring)))))",message="external mode enabled cluster cannot have monitoring field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.network)))))",message="external mode enabled cluster cannot have network field in CR"
+// +kubebuilder:validation:XValidation:rule="!has(self.external) || (has(self.external) && ((!has(self.external.enable) || self.external.enable == false )|| (self.external.enable == true && (!has(self.storage)))))",message="external mode enabled cluster cannot have storage field in CR"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:101]: Is there any way to simplify the statements when there are nullable fields? It would be nice if it just treated values of nullable fields as false or 0 if not set. This is a lot of complexity with nullable fields.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 5]
```diff
@@ -534,6 +542,8 @@ const (
 )
 
 // MonSpec represents the specification of the monitor
+// +kubebuilder:validation:XValidation:message="zones must be less than or equal to count",rule="!has(self.zones) || (has(self.zones) && (size(self.zones) <= self.count))"
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:545]: For stretch cluster we can verify specific values, which will simplify these rules as well.
- mon.count must be 5
- The number of zones must be exactly 3
- Exactly one zone should have `arbiter.true`

File: deploy/charts/rook-ceph/templates/resources.yaml [hunk 6]
```diff
@@ -2213,6 +2221,11 @@ spec:
                       type: object
                   type: object
                   x-kubernetes-preserve-unknown-fields: true
+                  x-kubernetes-validations:
```
Comment (travisn) [inline deploy/charts/rook-ceph/templates/resources.yaml:2224]: As a way to test these validations, it would be really nice to have a folder of examples that fail these checks. Then we can have an integration test that verifies they all fail. This could be a follow-up PR.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 4 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go:101]: No, I don't we can make it simpler `field presence can be checked via has(self.field)`. We have `   // +kubebuilder:validation:Optional` but will be valid for only one settings but not when we compare with multiple settings

File: deploy/charts/rook-ceph/templates/resources.yaml (hunk 6 above)
Comment (subhamkrai) [reply deploy/charts/rook-ceph/templates/resources.yaml:2224]: will open issue to track this.

Comment (subhamkrai) [reply deploy/charts/rook-ceph/templates/resources.yaml:2224]: https://github.com/rook/rook/issues/13229#issue-1999171755

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 5 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go:545]: This will be new rule, right?

File: deploy/examples/pre-1.26-crds.yaml
Comment (subhamkrai) [reply deploy/examples/pre-1.26-crds.yaml:1]: done

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 5 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:545]: These two rules about the mon count for stretch cluster don't make sense to me. I would suggest removing those two rules and adding the rules I mentioned above.

File: tests/framework/installer/ceph_manifests.go [hunk 7]
```diff
@@ -81,7 +85,18 @@ func (m *CephManifestsMaster) Settings() *TestCephSettings {
...
-func (m *CephManifestsMaster) GetCRDs(k8shelper *utils.K8sHelper) string {
+func (m *CephManifestsMaster) GetCRDs(t *testing.T, k8shelper *utils.K8sHelper) string {
+	k8sVersion, err := k8sutil.GetK8SVersion(k8shelper.Clientset)
+	if err != nil {
+		assert.NoError(t, err)
+	}
+	minValidatinPolicy := k8sVersion.AtLeast(version.MustParseSemantic(MinValidatinPolicy))
+	// For multi-cluster integration we create a internal mode cluster and external mode cluster both
+	// but since CRDs are not namespaced tight we need to use pre-1.25/crds.yaml so that CI works.
+	if !minValidatinPolicy || m.settings.ClusterName == "test-external" || m.settings.ClusterName == "multi-cluster" {
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go:96]: Isn't the crds.yaml version only dependent on the k8s version? I don't see why the external or multi-cluster tests would use the pre-1.25 crds. That would be an indication that external clusters are broken with the new CEL rules.
```suggestion
	if !minValidatinPolicy {
```

File: tests/framework/installer/ceph_manifests_previous.go [hunk 8]
```diff
@@ -37,8 +42,16 @@ func (m *CephManifestsPreviousVersion) Settings() *TestCephSettings {
 	return m.settings
 }
 
-func (m *CephManifestsPreviousVersion) GetCRDs(k8shelper *utils.K8sHelper) string {
-	return m.settings.readManifestFromGitHub("crds.yaml")
+func (m *CephManifestsPreviousVersion) GetCRDs(t *testing.T, k8shelper *utils.K8sHelper) string {
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests_previous.go:45]: The `CephManifestsPreviousVersion` type is only used for the upgrade tests. During the upgrade tests we always want to use the crds.yaml since the previous version (1.12) does not have the pre-1.25 crds. So the upgrade test should pass if you revert the changes to this file.

File: tests/framework/installer/ceph_manifests_previous.go [hunk 9]
```diff
@@ -37,8 +42,16 @@ func (m *CephManifestsPreviousVersion) Settings() *TestCephSettings {
...
+	k8sVersion, err := k8sutil.GetK8SVersion(k8shelper.Clientset)
+	if err != nil {
+		assert.NoError(t, err)
+	}
+	minValidatinPolicy := k8sVersion.AtLeast(version.MustParseSemantic(MinValidatinPolicy))
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests_previous.go:50]: typo, and seems like you can make it lower case if no other package uses it
```suggestion
	minValidatinPolicy := k8sVersion.AtLeast(version.MustParseSemantic(minValidationPolicy))
```

File: tests/framework/installer/ceph_manifests_previous.go [hunk 10]
```diff
@@ -37,8 +42,16 @@ func (m *CephManifestsPreviousVersion) Settings() *TestCephSettings {
...
+	if minValidatinPolicy {
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests_previous.go:51]: nit: I'd recommend it's more readable just to use this directly instead of use a variable.
```suggestion
	if k8sVersion.AtLeast(version.MustParseSemantic(minValidationPolicy) {
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by Madhu-1** [comment]
> Per discussion in slack, we will need to generate a separate crds.yaml to a subdirectory that will be used for different versions of K8s.
> 
> According to [this article](https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/), the validations graduated to beta in K8s 1.25, so I would hope that 1.25 and newer can use these. I would suggest we not enable any feature gates to get this working even in the CI. We should only recommend it be used without the feature gate enabled.

If the feature gate is not enabled the checks will be skipped in kubernetes, any specific r3ason to have different yaml files based on the kubernetes version?

**Comment by subhamkrai** [comment]
> 
> If the feature gate is not enabled the checks will be skipped in kubernetes, any specific r3ason to have different yaml files based on the kubernetes version?

@Madhu-1 The feature gate is disabled, the error we'll get is something similar to  `x-validation keyword unknown`. The crd schema will be invalid in the older version.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 5 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go:545]: the rules I have taken is from the existing webhook rules https://github.com/rook/rook/blob/510d2721b4ca350f26c998d528926eb1e3b66e4f/pkg/apis/ceph.rook.io/v1/cluster.go#L49-L61 PTAL and see if this rule and webhook matches and make sense.

What you suggesting is also doable but seems like different than what we already have in webhook checks

File: tests/framework/installer/ceph_manifests_previous.go (hunk 10 above)
Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests_previous.go]: this file reverted as this is used for previous version upgrades

File: tests/framework/installer/ceph_manifests.go (hunk 7 above)
Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests.go:96]: did you take time to read my comments just above this check
```text
// For multi-cluster integration we create a internal mode cluster and external mode cluster both
// but since CRDs are not namespaced tight we need to use pre-1.25/crds.yaml so that CI works.
```

Comment (BlaineEXE) [reply tests/framework/installer/ceph_manifests.go:96]: I thought we did some work to update other-namespace `common-second-cluster.yaml` recently. What about those?

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 5 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:545]: Oh I see now, this rule was intended for adding zones for non-stretch clusters in #12384. Let's keep this first rule for that case. But the stretch clusters always require exactly 3 zones. So the second rule (line 546) could check for exactly 3 zones.

File: tests/framework/installer/ceph_manifests.go (hunk 7 above)
Comment (travisn) [reply tests/framework/installer/ceph_manifests.go:96]: Correct, the CRDs are not namespaced. But the version of the CRDs used (CEL vs non-CEL) only depends on what version of K8s, so I still don't understand why the multi-cluster test would need to use the non-CEL version of the CRDs.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 5 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go:545]: added

File: tests/framework/installer/ceph_manifests.go (hunk 7 above)
Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests.go:96]: Let's discuss in huddle, but the issue happening, in once cluster where CRD with CEL rule is created. We are creating two cephCluster in different namespace, one is internal mode where we define `mons`, `osd`  and so and in second cluster we create external cluster, then the cel rule through the error because the cephcluster crd external mode can't have `mon` and `osd` field. So,

Comment (travisn) [reply tests/framework/installer/ceph_manifests.go:96]: This means the CEL rule is not ever allowing an external cluster at the same time as a converged cluster. Sounds like a bug in the CEL. We need to allow that.

Comment (travisn) [reply tests/framework/installer/ceph_manifests.go:96]: What was the error in the CI you were seeing? I was able to create both a converged cecphcluster CR and an external CephCluster CR in my local minikube cluster without hitting any errors from the CEL. And I don't see any issues in the external cluster yaml [created by the test](https://github.com/rook/rook/blob/580760c8d38a921620cd06836669724b59fec615/tests/framework/installer/ceph_manifests.go#L616-L630), so I would expect it to work.

Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests.go]: I removed this check and see the multiclustertest is failing and when I check the logs I can see 
```yaml
2023-11-29 05:50:24.396462 E | ceph-cluster-controller: failed to reconcile CephCluster "multi-external/test-external". failed to add finalizer: failed to add finalizer "cephcluster.ceph.rook.io" on "test-external": CephCluster.ceph.rook.io "test-external" is invalid: [spec: Invalid value: "object": external clusters cannot define the mon settings, spec: Invalid value: "object": external cluster cannot define dashboard settings, spec: Invalid value: "object": external cluster cannot define monitoring settings, spec: Invalid value: "object": external cluster cannot define network settings, spec: Invalid value: "object": external cluster cannot define storage settings, spec: Invalid value: "object": external cluster cannot define disruptionManagement settings, spec: Invalid value: "object": external cluster cannot define mgr settings]
```

Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests.go]: I was also able to create the external crd locally

Comment (travisn) [reply tests/framework/installer/ceph_manifests.go]: Actually I was able to see that error locally in the operator log, I had only been looking for errors when running the `kubectl create` command. When the operator tries to add the finalizer, some empty fields are causing the problem with the CEL of the external cluster. So this is really a problem for any external cluster, and not just a problem with the CI in that scenario with the multi-cluster test.

After I remove the CEL from the CRDs and allow the CR to be updated, this is what the spec looks like for the external CephCluster CR. So the CEL rules need to allow for the empty structs. And if it's not simple to get working, let's remove those CEL rules for the external cluster so we can get this merged, then we can follow up with a separate PR.
```yaml
  spec:
    cephVersion: {}
    cleanupPolicy:
      sanitizeDisks: {}
    crashCollector:
      disable: true
    dashboard: {}
    disruptionManagement: {}
    external:
      enable: true
    healthCheck:
      daemonHealth:
        mon:
          interval: 45s
        osd: {}
        status: {}
    logCollector: {}
    mgr: {}
    mon: {}
    monitoring: {}
    network:
      multiClusterService: {}
    security:
      keyRotation:
        enabled: false
      kms: {}
    storage:
      flappingRestartIntervalHours: 0
      store: {}
```
And let's make the change from the original comment:
```
if !minValidatinPolicy {
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 11]
```diff
@@ -2272,6 +2282,8 @@ type SSSDSidecarAdditionalFile struct {
 }
 
 // NetworkSpec for Ceph includes backward compatibility code
+// +kubebuilder:validation:XValidation:rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))",message="at least one network selector must be specified when using the multus"
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:2285]: Nit

```suggestion
// +kubebuilder:validation:XValidation:rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))",message="at least one network selector must be specified when using multus"
```

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 12]
```diff
@@ -2272,6 +2282,8 @@ type SSSDSidecarAdditionalFile struct {
...
+// +kubebuilder:validation:XValidation:rule="!has(self.provider) || (oldSelf.provider != 'host' || (oldSelf.provider == 'host' && self.provider != 'multus'))",message="invalid update: network provider from 'host' to 'multus' provider change is not allowed"
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:2286]: The NetworkProvider type isn't actually nullable since it's a string. This is a bug IMO. I think we can omit `!has(self.provider)` from the spec if we remove the incorrect `// +nullable` from the `Provider NetworkProviderType` line.

```suggestion
// +kubebuilder:validation:XValidation:rule="self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0)",message="at least one network selector must be specified when using the multus"
// +kubebuilder:validation:XValidation:rule="oldSelf.provider != 'host' || (oldSelf.provider == 'host' && self.provider != 'multus')",message="invalid update: network provider from 'host' to 'multus' provider change is not allowed"
```

Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:2286]: This isn't a limitation that I have specified. But I think maybe this is based on Travis's limitation that is codified here right now:
https://github.com/rook/rook/blob/3c43268d0a924c3966a253b1bf28a5d491cdedb1/pkg/apis/ceph.rook.io/v1/network.go#L78-L87

I think the equivalent rule is to allow provider to transition from `""` to `"host"` and from `"host"` to `""`, but not other transitions.

@travisn is that right?

Along that trend, I wonder if it makes sense broadly disallow network transitions from provider to provider. This would enforce that all transitions go from `"providerA"` to `""` to `"providerB"`. Thoughts?

E.g.
```
// +kubebuilder:validation:XValidation:message="network provider must be disabled (reverted to empty string) before a new provider is enabled",rule="(oldSelf.provider == '' && self.provider != '') || (oldSelf.provider != '' && self.provider == '')"
```

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go]: The conversation here doesn't answer the question I posed initially. Is it possible to reorder the items on each line so that the `message=` appears before `rule=`? 

I think this will make it easier to read at a glance what the rules' intentions are. Otherwise, we will only be reading the really arcane rule syntax and have to scroll an arbitrary distance to the right to read what it's supposed to do.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 12 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:2286]: Yes, transitioning to and from host should be allowed. Transitioning between other providers is not supported, or at least questionable. Sounds good if we have a simple rule that says not to transition between providers.

Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go]: added the change and tested

Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go]: we need to add `!has(self.provider)` check since we are evaluating it later in the rule and with this it will through error `unknow key ...` also it is not related to `// +nullable`

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 2 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go]: Yes, we can have `message=` appear first and I agree it will help in reading

File: tests/framework/installer/ceph_manifests.go [hunk 13]
```diff
@@ -81,7 +85,18 @@ func (m *CephManifestsMaster) Settings() *TestCephSettings {
...
+	k8sVersion, err := k8sutil.GetK8SVersion(k8shelper.Clientset)
+	if err != nil {
+		assert.NoError(t, err)
+	}
+	minValidatinPolicy := k8sVersion.AtLeast(version.MustParseSemantic(minValidatinPolicy))
+	// For multi-cluster integration we create a internal mode cluster and external mode cluster both
+	// but since CRDs are not namespaced tight we need to use pre-1.25/crds.yaml so that CI works.
+	if !minValidatinPolicy {
+		return m.settings.readManifest("pre-1.25/crds.yaml")
+
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go:98]: nit: remove blank line

File: deploy/examples/pre-1.25/crds.yaml
Comment (travisn) [inline deploy/examples/pre-1.25/crds.yaml:1]: How are we going to keep this file updated? If there are new settings added to type.go, we need those to be generated here. The intention of this version of the crds.yaml is just to remove the validations from the schema.

Comment (travisn) [reply deploy/examples/pre-1.25/crds.yaml:1]: Actually, are these validation properties just ignored on older versions of K8s? They must be, or else the helm tests would be failing on K8s 1.23 since the helm chart always creates the rules. In that case, it seems we don't need this pre-1.25 version of the schema? Or else I could be forgetting the reason we needed to create the separate version.

**Review by travisn — CHANGES_REQUESTED** [review]

File: deploy/examples/pre-1.25/crds.yaml
Comment (subhamkrai) [reply deploy/examples/pre-1.25/crds.yaml:1]: Yes, the reason we split is due to older versions of k8s.  If clulsters/users running older version of k8s they will use this `pre-1.25/crds.yaml` and IMO we don't need to update or maintain that like how we did for `v1beta1` crds in the past. We'll only update the `crds.yaml` which has CEL rules.

Comment (subhamkrai) [reply deploy/examples/pre-1.25/crds.yaml:1]: if you see this ci https://github.com/rook/rook/actions/runs/6847034856/job/18614595031#step:5:68 we got this error for smoke suite 
```bash
error: error validating "STDIN": error validating data: [ValidationError(CustomResourceDefinition.spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.dataDirHostPath): unknown field "x-kubernetes-validations" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps, ValidationError(CustomResourceDefinition.spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.mon): unknown field "x-kubernetes-validations" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps, ValidationError(CustomResourceDefinition.spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.network): unknown field "x-kubernetes-validations" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps, ValidationError(CustomResourceDefinition.spec.versions[0].schema.openAPIV3Schema.properties.spec): unknown field "x-kubernetes-validations" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps]; if you choose to ignore these errors, turn validation off with --validate=false
```

Comment (travisn) [reply deploy/examples/pre-1.25/crds.yaml:1]: It seems that was caused by K8s 1.22, but seems ok in 1.23 which would explain why the helm test is passing. Let's try removing the pre-1.25 crds to see if we can actually keep a single crds file.

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 12 above)
Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go]: I didn't get the error when I can the `kubectl create -f cluster-multus-test.yaml` but in the operator logs I'm seeing the error.

Comment (subhamkrai) [reply pkg/apis/ceph.rook.io/v1/types.go]: well came with more simpler rule to understand `rule="self == '' || self == oldSelf"` and tested it as well with watching operator logs, worked

File: tests/framework/installer/ceph_manifests.go [hunk 14]
```diff
@@ -81,7 +82,17 @@ func (m *CephManifestsMaster) Settings() *TestCephSettings {
...
+func (m *CephManifestsMaster) GetCRDs(t *testing.T, k8shelper *utils.K8sHelper) string {
+	// k8sVersion, err := k8sutil.GetK8SVersion(k8shelper.Clientset)
+	// if err != nil {
+	// 	assert.NoError(t, err)
+	// }
+	// minValidatinPolicy := k8sVersion.AtLeast(version.MustParseSemantic(minValidatinPolicy))
+	// // For multi-cluster integration we create a internal mode cluster and external mode cluster both
+	// // but since CRDs are not namespaced tight we need to use pre-1.25/crds.yaml so that CI works.
+	// if !minValidatinPolicy {
+	// 	return m.settings.readManifest("pre-1.25/crds.yaml")
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go:94]: Great to see the CI is working now without a different version of the CRDs! So we can revert the pre-1.25 crds.yaml, right?

Comment (subhamkrai) [reply tests/framework/installer/ceph_manifests.go]: reverted

File: deploy/examples/operator.yaml [hunk 15]
```diff
@@ -503,7 +503,7 @@ data:
   # The CSI GRPC timeout value (in seconds). It should be >= 120. If this variable is not set or is an invalid value, it's default to 150.
   CSI_GRPC_TIMEOUT_SECONDS: "150"
 
-  ROOK_DISABLE_ADMISSION_CONTROLLER: "true"
+  ROOK_DISABLE_ADMISSION_CONTROLLER: "false"
```
Comment (travisn) [inline deploy/examples/operator.yaml:506]: revert?

File: tests/framework/installer/ceph_manifests_previous.go [hunk 16]
```diff
@@ -18,13 +18,15 @@ package installer
...
+	"testing"
 
 	"github.com/rook/rook/tests/framework/utils"
 )
 
 const (
 	// The version from which the upgrade test will start
-	Version1_12 = "v1.12.7"
+	Version1_12        = "v1.12.7"
+	minValidatinPolicy = "1.25.0"
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests_previous.go:29]: not needed?

File: tests/framework/installer/ceph_manifests.go [hunk 17]
```diff
@@ -81,7 +82,7 @@ func (m *CephManifestsMaster) Settings() *TestCephSettings {
 	return m.settings
 }
 
-func (m *CephManifestsMaster) GetCRDs(k8shelper *utils.K8sHelper) string {
+func (m *CephManifestsMaster) GetCRDs(t *testing.T, k8shelper *utils.K8sHelper) string {
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests.go:85]: This is no longer needed either?

**Review by travisn — APPROVED** [review]

**Review by BlaineEXE — APPROVED** [review]
